src/core/stream/representation/downloading_queue.ts

```diff
@@ -0,0 +1,510 @@
+   * load segments for.
+   * @param {BehaviorSubject} downloadQueue$ - Emit the queue of segments you
+   * want to load.
+   * @param {Object} segmentFetcher - Interface to facilitate the download of
+   * segments.
+   * @param {boolean} hasInitSegment - Declare that an initialization segment
+   * will need to be downloaded.
+   *
+   * A `DownloadingQueue` ALWAYS wait for the initialization segment to be
+   * loaded and parsed before parsing a media segment.
+   *
+   * In cases where no initialization segment exist, this would lead to the
+   * `DownloadingQueue` waiting indefinitely for it.
+   *
+   * By setting that value to `false`, you anounce to the `DownloadingQueue`
+   * that it should not wait for an initialization segment before parsing a
+   * media segment.
+   */
+  constructor(
+    content: IDownloadingQueueContext,
+    downloadQueue$ : BehaviorSubject<IDownloadQueueItem>,
+    segmentFetcher : IPrioritizedSegmentFetcher<T>,
+    hasInitSegment : boolean
+  ) {
+    this._content = content;
+    this._currentObs$ = null;
+    this._downloadQueue$ = downloadQueue$;
+    this._initSegmentRequest = null;
+    this._mediaSegmentRequest = null;
+    this._segmentFetcher = segmentFetcher;
+    this._initSegmentMetadata$ = new ReplaySubject<number|undefined>(1);
+    this._mediaSegmentsAwaitingInitMetadata = new Set();
+    if (!hasInitSegment) {
+      this._initSegmentMetadata$.next(undefined);
+    }
+  }
+
+  /**
+   * Returns the initialization segment currently being requested.
+   * Returns `null` if no initialization segment request is pending.
+   * @returns {Object}
+   */
+  public getRequestedInitSegment() : ISegment | null {
+    return this._initSegmentRequest === null ? null :
+                                               this._initSegmentRequest.segment;
+  }
+
+  /**
+   * Returns the media segment currently being requested.
+   * Returns `null` if no media segment request is pending.
+   * @returns {Object}
+   */
+  public getRequestedMediaSegment() : ISegment | null {
+    return this._mediaSegmentRequest === null ? null :
+                                                this._mediaSegmentRequest.segment;
+  }
+
+  /**
+   * Start the current downloading queue, emitting events as it loads and parses
+   * initialization and media segments.
+   *
+   * If it was already started, returns the same - shared - Observable.
+   * @returns {Observable}
+   */
+  public start() : Observable<IDownloadingQueueEvent<T>> {
+    if (this._currentObs$ !== null) {
+      return this._currentObs$;
+    }
+    const obs = observableDefer(() => {
+      const mediaQueue$ = this._downloadQueue$.pipe(
+        filter(({ segmentQueue }) => {
+          // First, the first elements of the segmentQueue might be already
+          // loaded but awaiting the initialization segment to be parsed.
+          // Filter those out.
+          let nextSegmentToLoadIdx = 0;
+          while (true) {
```

Comment:
Just out of curiosity why would an Infinite loop `while (true)` is better than a `for` only on the element that is presented in the array `segmentQueue`?

Do u get some inspiration from Rust where a loop like this is more common?

Reply:
It was a long time since I wrote that but I think I write a `while (true) {` instead of a `while (condition)` because conditions were multiple and one depended on the other + we needed to define `nextSegment`.
With a `for`, I don't know how it could be written legibly.

Even if that's true that the equivalent `loop` construct is more common in Rust I don't think that I wanted to transpose that idiom here. Anyway, we should keep writing idiomatic JS code regardless of what people do in other languages.

If you think of a more readable way to write that code I'm open.

Reply:
It could be written like that no ?

```js
let nextSegmentToLoadIdx = 0;
for (let i = 0; i < segmentQueue.length; i++) {
  const nextSegment = segmentQueue[nextSegmentToLoadIdx].segment;
  if (this._mediaSegmentsAwaitingInitMetadata.has(nextSegment.id)) {
     nextSegmentToLoadIdx++;
   } else {
      break;
   }
}
```

But anyway, it doesn't bother me as much as It sounds, so feel free to keep it like this, if u feel it :)

Reply:
you're right that it's more readable like that.

Just, there is the `i` variable which seems unneeded and confusing because it should stay implicitely always equal to `nextSegmentToLoadIdx`.

Maybe I should write something more like:
```
for (; nextSegmentToLoadIdx < segmentQueue.length; nextSegmentToLoadIdx++) {
```

I'm on it

Reply:
done